### PR TITLE
Add output file feature for ssl_client2.c

### PR DESCRIPTION
**Pull request by ssrlive**
Add a `output file` option for the `ssl_client2.c` module.

The purpose of adding this feature is to make it easier for testers to use this client to download files directly from the https website. Of course, there are a lot of condition judgements  are not implemented  in this commits, but it is enought for working in most cases.

### CODE REVIEW

**Comment by hanno-arm** [comment]
Hi @ssrlive,

thank you for your contribution! Could you please provide a short description of the motivation for and approach behind this PR? This will help to speed up the review process.

Kind regards,
Hanno, Mbed TLS team member

**Comment by ssrlive** [comment]
I have updated the description.

**Comment by RonEld** [comment]
@ssrlive Thank you for your contribution!
In addition to what @hanno-arm has asked, unfortunately our policy is to not accept contributions, without a Contributor’s Licence Agreement (CLA) signed or authorised by yourself or your employer.

You can find the [agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us.

Thanks for your understanding and again, thanks for the contribution!

**Comment by RonEld** [comment]
CI failed on the following line:
~~~
ssl/ssl_client2.c: In function 'main':

ssl/ssl_client2.c:1930:33: error: unknown conversion type character 'z' in format [-Werror=format=]

                 mbedtls_printf( " %zu bytes read\n", len0 );
~~~

In addition, note that `ssl_client2` is an example reference program, and should not cover all cases.

Please add to the description justification for this enhancement, meaning, why do you think this feature is necessary?

**Comment by ssrlive** [comment]
@RonEld  Well. Since I want to apply mbedTLS to my project [ShadowsocksR-Native](https://github.com/ShadowsocksR-Live/shadowsocksr-native) , So I use `ssl_client2.c` to test mbedTLS to download functionalilty with the https protocol.
But I found that i always fails while download files larger than 1mb with ssl_client2.c .
So I want to add the 'output file' feature to this tool, so that you guys and others can do further testing and bug fixes.

**Comment by ssrlive** [comment]
pleas review the code.

File: programs/ssl/ssl_client2.c [hunk 1]
```diff
@@ -116,7 +116,7 @@ int main( void )
 #define DFL_EXTENDED_MS         -1
 #define DFL_ETM                 -1
 
-#define GET_REQUEST "GET %s HTTP/1.0\r\nExtra-header: "
+#define GET_REQUEST "GET %s HTTP/1.0\r\nHost: %s\r\nConnection: keep-alive\r\nExtra-header: "
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: How is changing the headers related to adding the output file feature? This seems like an unrelated change and deserves its own commit message with its own justification written in the commit message.

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -594,7 +597,7 @@ int main( int argc, char *argv[] )
     memset( (void * ) alpn_list, 0, sizeof( alpn_list ) );
 #endif
 
-    if( argc == 0 )
+    if( argc <= 1 )
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: Do we require that the output file be specified? If not, why do we need to change the minimum required number of arguments?

File: programs/ssl/ssl_client2.c [hunk 3]
```diff
@@ -1728,7 +1737,7 @@ int main( int argc, char *argv[] )
     fflush( stdout );
 
     len = mbedtls_snprintf( (char *) buf, sizeof( buf ) - 1, GET_REQUEST,
-                            opt.request_page );
+                            opt.request_page, opt.server_name );
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: When we move the header change to a separate commit, this `mbedtls_snprintf()` change needs to move to that same commit.

File: programs/ssl/ssl_client2.c (hunk 1 above)
Comment (ssrlive) [reply programs/ssl/ssl_client2.c]: If the request header is missing the host field, many http servers will refuse to return the resource you requested, causing the request to fail. So I added it.

File: programs/ssl/ssl_client2.c (hunk 2 above)
Comment (ssrlive) [reply programs/ssl/ssl_client2.c]: The original value of argc is always greater than 0, because the first argument passed to argv[] is always the path to the program itself.
I don't understand why you determine if it is equal to 0, which makes no sense.

Comment (Patater) [reply programs/ssl/ssl_client2.c]: OK, this is another fix that should go in its own commit. So far this PR looks like it fixes 2 issues (header, argc) and also adds a new feature; I'd expect 3 commits.

Comment (ssrlive) [reply programs/ssl/ssl_client2.c]: OK. As your wish. code changed.

**Comment by ssrlive** [comment]
> The commit organization here is not up to our standards. Please follow https://os.mbed.com/docs/mbed-os/v5.10/reference/workflow.html#guidelines-for-github-pull-requests, ensuring you've adequately explained why this change is necessary in the commit message body. Please also make unrelated changes in separate commits.

Have changed. Please merge it.

**Comment by mpg** [comment]
The 2.16 branch as now reached and of life, so I'm closing this PR. Feel free to open a new PR based on the `development` branch. Thanks for you contribution.
